File: Sheets/src/neetcode/practice/Arrays/_9_longestConsecutiveSequence.java
package neetcode.practice.Arrays;

import java.util.HashSet;

class Solution {
  public int longestConsecutive(int[] nums) {
    HashSet<Integer> uniqueElements = new HashSet<>() {
      {
        for (int i : nums)
          this.add(i);
      }
    };

    int longestLength = 0, currentLength = 1;
    for (int i = 0; i < nums.length; i++) {
      int num = nums[i];
      if (!uniqueElements.contains(num - 1)) {
        while (true) {
          if (uniqueElements.contains(++num))
            currentLength++;
          else
            break;
        }
      }
      longestLength = Math.max(currentLength, longestLength);
      currentLength = 1;
    }

    return longestLength;
  }
}

File: Sheets/src/neetcode/practice/Arrays/_9_LongestConsecutiveSequence.cc
#include <algorithm>
#include <vector>

class Solution {
public:
  int longestConsecutive(std::vector<int> &nums) {
    ptrdiff_t maxLength = 0, left = 0, n = nums.size();
    std::sort(nums.begin(), nums.end());
    nums.erase(std::unique(nums.begin(), nums.end()), nums.end());

    bool inSeq = false;
    for (ptrdiff_t i = 1; i < n; i++) {
      if (inSeq) {
        if (nums[i] - nums[i - 1] != 1) {
          maxLength = std::max(maxLength, i - left);
          left = i;
          inSeq = false;
        }
      } else {
        if (nums[i] - nums[i - 1] == 1) {
          inSeq = true;
          continue;
        }
        left = i;
      }
    }

    return static_cast<int>(std::max(maxLength, n - left));
  }
};
